students/eowyn/session09/oo-mailroom/oomailroom.py

```diff
@@ -0,0 +1,232 @@
+
+    @property
+    def average_donation(self):
+        return round(self.sum_donations / self.count_donations, 2)
+
+
+class Transactions():
+    # Collect list of Donor objects and generate reports
+
+    def __init__(self):
+        self.all_donors = []
+
+    def add_donor(self, name, amt):
+        this_donor = self.get_donor(name)
+        if this_donor:
+            this_donor.add_donation(amt)
+        else:
+            this_donor = Donor(name)
+            this_donor.add_donation(amt)
+            self.all_donors.append(this_donor)
+        return this_donor.generate_letter()
+
+    def get_donor(self, name):
+        for i in self.all_donors:
+            if i.name == name:
+                return i
+
+    def list_names(self):
+        return ("Donor Names:\n" +
+                '\n'.join([i.name for i in self.all_donors]))
+
+    def generate_report_data(self):
+        names = [i.name for i in self.all_donors]
+        totals = [i.sum_donations for i in self.all_donors]
+        counts = [i.count_donations for i in self.all_donors]
+        averages = [i.average_donation for i in self.all_donors]
+        report_data = list(zip(names, totals, counts, averages))
+        report_data = sorted(report_data,
+                             key=lambda y: int(y[1]),
+                             reverse=True)
+        return report_data
+
+    def setup_table(self):
+        ''' Set up the donor record table headers '''
+        headers = ('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+        fs1 = '|'.join(["{:<40}", "{:<12}", "{:<10}", "{:<12}"])
+        width = len(fs1.format(*headers))
+        head = (fs1.format(*headers)) + '\n'
+        head = head + (width * "-") + '\n'
+        return head
+
+    def setup_body(self):
+        ''' Set up the donor records table body '''
+        body = ""
+        fs2 = ''.join(["{:<40}", "${:^12.2f}", "{:^12d}", "${:^12.2f}"])
+        list_data = self.generate_report_data()
+        for i in range(len(list_data)):
+            entry = list_data[i]
+            body = body + fs2.format(*entry) + '\n'
+        return body
+
+    def print_donor_records(self):
+        ''' Pretty-print the donor records '''
+        print(self.setup_table())
+        print(self.setup_body())
+
+
+class Mailroom():
+
+    # --------------------------------------------------------------
+    # Following are helper functions to control program flow
+    # -------------------------------------------------------------------
+
+    def safe_input(self):
+        return None
+
+    def quit_code(self):
+        sys.exit()
+
+    def return_to_menu(self):
+        ''' Return True to trigger exit out of sub-loop'''
+        return True
+
+    # --------------------------------------------------------------
+    # Following are helper functions for accepting and responding to
+    # keyboard input
+    # ----------------------------------------------------------------
+
+    def __init__(self, transactions):
```

Comment:
you usually put your __init__ at the top of the class -- that way you can see right away how to initialize the class.